packages/table/src/useTable.tsx

```diff
@@ -0,0 +1,210 @@
+import * as React from 'react'
+import { SxStyleProp } from 'theme-ui'
+import { useId } from '@kodiak-ui/hooks'
+import { hasKey } from './utils'
+import { useRowSelect, createRowState, SelectedRowsState } from './useRowSelect'
+
+export type ID = string | number
+export type Scope = 'col' | 'row'
+export type TableLayout = 'auto' | 'fixed'
+
+export type Data = {
+  id: ID
+}
+
+export type Column = {
+  Cell: string | React.ReactNode
+  accessor?: string
+  scope?: Scope
+  width?: string
+}
+
+export type Cell = {
+  column?: Column
+  key: string
+  children: React.ReactNode
+  rowData?: any
+  width?: string
+}
+
+export type Header = { scope?: Scope } & Cell
+
+export type Row = {
+  key: string
+  id: ID
+  cells: Cell[]
+  rowData: any
+}
+
+export interface UseTableProps<T> {
+  columns: Column[]
+  data: T[]
+  tableLayout?: TableLayout
+  selectable?: boolean
+  initialSelectedIds?: ID[]
+  attributes?: {
+    id?: string
+    describedby?: React.MutableRefObject<any> | string
+  }
+}
+
+export type GetTableProps = {
+  id: string
+  sx: SxStyleProp
+  'aria-describedby'?: string
+}
+
+export interface UseTableReturnValue {
+  headers: Header[]
+  rows: Row[]
+  selectedRows: SelectedRowsState
+  flatSelectedRows: ID[]
+  allSelected: boolean
+  someSelected: boolean
+  selectedCount: number
+  getTableProps: () => GetTableProps
+}
+
+export function useTable<T extends Data>({
+  columns,
+  data,
+  tableLayout = 'auto',
+  selectable = false,
+  initialSelectedIds,
+  attributes = {},
+}: UseTableProps<T>): UseTableReturnValue {
+  const { id: userId, describedby } = attributes
+  const id = useId(userId)
+  const {
+    selectedRows,
+    flatSelectedRows,
+    allSelected,
+    someSelected,
+    selectedCount,
+    setSelectedRows,
+    getSelectAllCheckbox,
+    getRowCheckbox,
+  } = useRowSelect({
+    selectable,
+    initialState: createRowState<T>({ data, state: false, initialSelectedIds }),
+    onSelect: (event, id) =>
+      setSelectedRows({
+        ...selectedRows,
+        [id]: event.target.checked,
+      }),
+    onSelectAll: event =>
```

Comment:
Do we need a way to allow this to be overridden for a case when we want to only select the items on the page vs all of the items?

Reply:
Yea, we will. Let's think through that scenario a bit more.